File: message/message-java/src/main/java/quan/message/MessageFactory.java
package quan.message;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import quan.common.util.ClassUtils;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * 消息工厂
 * Created by quanchangnai on 2019/6/20.
 */
public class MessageFactory {

    protected final Logger logger = LoggerFactory.getLogger(getClass());

    protected Map<Integer, Message> prototypes = new HashMap<>();

    public void register(Message message) {
        if (prototypes.containsKey(message.getId())) {
            throw new RuntimeException("消息ID重复:" + message.getId());
        }
        prototypes.put(message.getId(), message);
    }

    public void autoRegister(String packageName) {
        Set<Class<?>> messageClasses = ClassUtils.loadClasses(packageName, Message.class);
        for (Class<?> messageClass : messageClasses) {
            try {
                Message message = (Message) messageClass.getDeclaredConstructor().newInstance();
                register(message);
            } catch (Exception e) {
                logger.error("", e);
            }
        }
    }

    public Message create(int msgId) {
        Message message = prototypes.get(msgId);
        if (message == null) {
            return null;
        }
        return message.create();
    }

}

File: quan.lua/test/src/test/message/MessageFactory.lua
---
---消息工厂
---自动生成
---
local MessageFactory = {}

local prototypes = {
    [544233] = "test.message.role.CRoleLogin",
    [763075] = "test.message.role.SRoleLogin",
}

---
---消息工厂.创建消息
---@param msgId 消息ID
---
function MessageFactory.create(msgId)
    assert(math.type(msgId) == "integer", "参数[msgId]类型错误")
    return require(prototypes[msgId]).new()
end

return MessageFactory